tinygraph/tinygraph-rng.c

```diff
@@ -16,6 +22,184 @@
  *   Section 6.3.4 "PCG-RXS-M-XS"
  */
 
+#ifdef __AVX2__
+
+
+typedef struct tinygraph_rng {
+  TINYGRAPH_ALIGN(32) __m256i state;
+  TINYGRAPH_ALIGN(32) __m256i mul;
+  TINYGRAPH_ALIGN(32) __m256i inc;
+  TINYGRAPH_ALIGN(32) uint32_t cache[8];
+  uint32_t cache_idx;
+} tinygraph_rng;
```

Comment:
The `state` holds 8x uint32 pcg state; `mul` is the constant multiplier that's the same for all pcgs, `inc` is the increment that's different (and must be odd) for every single pcg, and `cache` and `cache_idx` are so that we can batch-generate 8 random values with avx2 and still have a function returning a single value to the user.